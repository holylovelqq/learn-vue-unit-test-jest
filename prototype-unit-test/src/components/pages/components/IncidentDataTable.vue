<template>
  <div class="incident-data-table">
    <el-table
      :data="incidents"
      :default-sort = "{prop: 'id', order: 'ascending'}"
      border
      empty-text = "No Data"
      style="width: 100%">
      <el-table-column
        class="eltablecolum"
        prop="id"
        label="ID"
        sortable
        width="80">
      </el-table-column>
      <el-table-column
        class="eltablecolum"
        prop="title"
        label="タイトル"
        sortable
        width="">
        <template slot-scope="scope">
          <el-button @click="goIncidentDetail(scope.row.id)" type="text">{{ scope.row.title }}</el-button>
        </template>
      </el-table-column>
      <!-- <el-table-column
        prop="affected_to"
        label="影響アプリケーション"
        sortable
        width="240">
      </el-table-column> -->
      <el-table-column
        class="eltablecolum"
        :formatter='formatOccurred_at'
        prop="occurred_at"
        label="発生日時"
        sortable
        width="180">
      </el-table-column>
      <el-table-column
        class="eltablecolum"
        prop="assigned_to"
        label="担当者"
        sortable
        width="180">
      </el-table-column>
    </el-table>
  </div>
</template>

<script>
export default {
  name: 'incident-data-table',
  props: {
    incidents: Array
  },
  methods: {
    goIncidentDetail: function (incidentId) {
      this.$router.push({
        name: 'IncidentDetail',
        params: {
          id: incidentId
        }
      })
    },
    formatOccurred_at: function (date) {
      return this.formatDatetime(date.occurred_at)
    },
    formatDatetime: function (date) {
      var m = this.$moment.utc(date, 'YYYY-MM-DDTHH:mm:ss.SSSZ')
      // タイムゾーンは分で指定する
      // 日本：9*60=54min
      // https://stackoverflow.com/questions/32878197/updating-time-offset-with-moment-utcoffset
      return m.utcOffset(0).format('YYYY/MM/DD HH:mm')
    }
  }
}
</script>
