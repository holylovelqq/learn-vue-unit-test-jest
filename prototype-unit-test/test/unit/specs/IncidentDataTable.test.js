import { shallowMount, createLocalVue } from '@vue/test-utils'
import elementUI from 'element-ui'
import moment from 'moment'
import IncidentDataTable from '@/components/pages/components/IncidentDataTable'

let $router = []
const localVue = createLocalVue()
localVue.use(elementUI)

localVue.prototype.$moment = moment

describe('IncidentDataTable.vue', () => {
  let incidents = [
    {
      'assigned_to': '山田太郎',
      'id': 1,
      'occurred_at': '2018-10-16T02:30:42.362000Z',
      'title': 'もんだい1'
    },
    {
      'assigned_to': '山田太郎2',
      'id': 2,
      'occurred_at': '2018-10-16T02:30:42.362000Z',
      'title': 'もんだい2'
    },
    {
      'assigned_to': '山田太郎3',
      'id': 3,
      'occurred_at': '2018-10-16T02:30:42.362000Z',
      'title': 'もんだい3'
    }
  ]
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(IncidentDataTable, {
      propsData: {incidents},
      localVue,
      mocks: {
        $router
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })
  it('IncidentDataTable DOM test', () => {
    expect(wrapper.is(IncidentDataTable)).toBeTruthy()
    expect(wrapper.find('.incident-data-table').exists()).toBe(true)
    expect(wrapper.findAll('.eltablecolum')).toHaveLength(4)
  })
  it('IncidentDataTable function test', () => {
    expect(wrapper.vm.formatOccurred_at({'occurred_at': '2018-10-16T02:30:42.362000Z'})).toBe('2018/10/16 02:30')
    wrapper.vm.$router = []
    const $r = [
      {
        name: 'IncidentDetail',
        params: {
          id: 5
        }
      }
    ]
    wrapper.vm.goIncidentDetail(5)
    expect(wrapper.vm.$router).toEqual($r)
  })

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
