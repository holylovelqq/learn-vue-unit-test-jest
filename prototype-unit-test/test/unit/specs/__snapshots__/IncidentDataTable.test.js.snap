// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncidentDataTable.vue matches snapshot 1`] = `
<div class="incident-data-table">
    <eltable-stub data="[object Object],[object Object],[object Object]" fit="true" border="true" showheader="true" emptytext="No Data" defaultsort="[object Object]" selectonindeterminate="true" style="width: 100%;">
        <eltablecolumn-stub type="default" label="ID" prop="id" width="80" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending," class="eltablecolum"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="タイトル" prop="title" width="" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending," class="eltablecolum"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="発生日時" prop="occurred_at" width="180" sortable="" resizable="true" formatter="function () { [native code] }" filtermultiple="true" sortorders="ascending,descending," class="eltablecolum"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="担当者" prop="assigned_to" width="180" sortable="" resizable="true" filtermultiple="true" sortorders="ascending,descending," class="eltablecolum"></eltablecolumn-stub>
    </eltable-stub>
</div>
`;
